Introduce profile for environment specific configuration

mowi12 opened the pull request:
Closes #218 

## What I have made

- Renamed `USE_DUMMY_USER` to `AUTH_BYPASS`
- Added `DATABASE_SEED_USER` env var for inserting a default user into the db
- Added `PROFILE` env var for testing, development and production profiles

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 commented:
In GitLab by @dts84 on Apr 22, 2022, 08:06

mentioned in commit c29cf95173aaf598a9eaa5d1b7072ad604f2a725

mowi12 committed:
feat: introduce application profile enum

mowi12 committed:
feat: Refactor USE_DUMMY_USER and add SEED_USER

Renamed the `USE_DUMMY_USER` var to `BYPASS_AUTHENTICATION` as the previous logic did exactly that.
Added a new `SEED_USER` var which inserts a default user into the database.

mowi12 committed:
docs: Update documentation

mowi12 committed:
feat: Update .env.example

luca-schlecker commented on wiki/Getting-Started.md:
```diff
@@ -25,29 +25,49 @@ environment variable (default: `8081`).
-project. Either create the file manually or copy the provided example:
+The app's configuration is driven by the `PROFILE` environment variable, which sets sensible defaults for different
+environments. You can set them in a `.env` file in the root directory. Either create the file manually or copy the
+provided example:
```
The wording is a bit weird. You talk about the `PROFILE` variable, yet you continue in plural without introducing the fact, that you can still control all the other variables.

luca-schlecker reviewed (changes requested):
This is looking great! Only one sentence could use some work. 👍

mowi12 reviewed (commented):
Thanks for the review. I applied the new changes!

But there was a small mistake. @Slartibartfass2 reviewed this PR as well, but in the #222 since this is based on this PR.
I will include his requests as well. Since he already openend the conversations in the other PR, they will be answered there.
So you both will review this PR : )

mowi12 committed:
refactor: extract logic to EnvService for better readability

luca-schlecker commented on src/main/kotlin/se/uulm/snowballr/backend/env/EnvReader.kt:
```diff
@@ -77,8 +82,8 @@ class EnvReader(
-        val port = defaultPort?.let { envService.getOrDefault(PORT, it.toString()).toInt() } ?: envService[PORT].toInt()
-        val host = defaultDBHost?.let { envService.getOrDefault(DATABASE_HOST, it) } ?: envService[DATABASE_HOST]
+        val port = envService.getRequiredOrDefault(PORT, defaultPort?.toString()).toInt()
+        val host = envService.getRequiredOrDefault(DATABASE_HOST, defaultDBHost)
```
This is more readable than what you had previously. 👍

luca-schlecker commented on wiki/Getting-Started.md:
```diff
@@ -64,9 +64,11 @@ The `PROFILE` variable determines the default behavior of the application.
+|---------------|-----------|-------------|-----------|
+| `PRODUCTION`  | `INFO`    | `false`     | `false`   |
+| `DEVELOPMENT` | `DEBUG`   | `false`     | `true`    |
+| `TESTING`     | `TRACE`   | `true`      | `true`    |
```
This is also great!

luca-schlecker reviewed (dismissed):
This is great! Props to @Slartibartfass2 for coming up with these ideas. I thought it was fine before the changes, but this such an improvement. LGTM 🚀 , but you can also wait for @Slartibartfass2 just in case.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/Main.kt:
```diff
@@ -36,7 +36,7 @@ fun main() {
  * @param logLevel The desired log level for the root logger. Acceptable values are one of
  * [io.github.oshai.kotlinlogging.Level] as string such as `DEBUG` or `INFO`.
- * If an invalid log level is provided, the default log level [DEFAULT_LOG_LEVEL] will be used.
+ * If an invalid log level is provided, the `DEBUG` log level will be used.
```
This needs to be undone again.

Slartibartfass2 commented on wiki/Getting-Started.md:
```diff
@@ -25,29 +25,51 @@ environment variable (default: `8081`).
 
+#### Profile-based Defaults
+
+| Profile       | LOG_LEVEL | AUTH_BYPASS | SEED_USER |
```
```suggestion
| Profile       | `LOG_LEVEL` | `AUTH_BYPASS` | `SEED_USER` |
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/env/EnvService.kt:
```diff
@@ -54,6 +60,12 @@ class EnvService : IEnvService {
+        return default?.let { getOrDefault(key, it) }
+            ?: dotenv[key]
+            ?: throw EnvVariableNotFoundException(key)
+    }
```
I like this, but it's a bit difficult to read. What about splitting this expression into two parts?

```suggestion
    override fun getRequiredOrDefault(key: String, default: String?): String {
        if (default != null) {
            return getOrDefault(key, default)
        }

        return dotenv[key] ?: throw EnvVariableNotFoundException(key)
    }
```

Slartibartfass2 reviewed (changes requested):
I like it. I commented anew on this PR so that the conversations are not in the other one.

mowi12 reviewed (commented):
All done! Only the stuff in the other PR left.

mowi12 committed:
refactor: Add '_ENABLED' suffix to boolean env vars

mowi12 committed:
refactor: reduce EnvReader complexity of `init` block

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/db/Database.kt:
```diff
@@ -126,38 +126,41 @@ class Database(
-    fun setupDummyUser() {
-        val useDummyUser = envReader.env.miscellaneous.useDummyUser
+    fun seedDummyUserIfEnabled() {
+        val seeUserEnabled = envReader.env.database.seedUserEnabled
```
```suggestion
        val seedUserEnabled = envReader.env.database.seedUserEnabled
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/db/Database.kt:
```diff
@@ -126,38 +126,41 @@ class Database(
-            // If the dummy user is not needed, delete it if it exists
-            UserTable.deleteWhere { UserTable.id eq existingId }.also {
-                logger.info { "Dummy user deleted" }
+        if (seeUserEnabled) {
```
```suggestion
        if (seedUserEnabled) {
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/db/Database.kt:
```diff
@@ -126,38 +126,41 @@ class Database(
-     * This method checks if a dummy user is needed based on the environment settings. If so, it
-     * creates the dummy user if it does not already exist. If the dummy user is not needed, it
-     * deletes the dummy user from the database, in case it exists.
+     * This method checks the `seedUser` flag. If true, it ensures the dummy user exists.
```
```suggestion
     * This method checks the `seedUserEnabled` flag. If true, it ensures the dummy user exists.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/grpc/interceptor/AuthenticationInterceptor.kt:
```diff
@@ -52,11 +52,12 @@ private val PUBLIC_METHODS =
+ * A whitelist of gRPC methods that are allowed to be called when authentication bypass is active.
  *
- * Only these methods are allowed to be called by the dummy user.
+ * When `AUTH_BYPASS` is enabled, only calls to methods listed in this set will be permitted
```
```suggestion
 * When `AUTH_BYPASS_ENABLED` is true, only calls to methods listed in this set will be permitted
```

Slartibartfass2 commented on wiki/Getting-Started.md:
```diff
@@ -25,29 +25,53 @@ environment variable (default: `8081`).
+| Profile               | `PRODUCTION` | `DEVELOPMENT` | `TESTING`   |
+|-----------------------|--------------|---------------|-------------|
+| `PORT`                | -            | 8080          | 8080        |
+| `DATABASE_HOST`       | `false`      | `localhost`   | `localhost` |
```
```suggestion
| `DATABASE_HOST`       | -            | `localhost`   | `localhost` |
```

mowi12 replied:
Copy-Pasta ...

Slartibartfass2 commented on wiki/Testing.md:
```diff
@@ -186,23 +189,39 @@ See
+
+### `AUTH_BYPASS_ENABLED`
+
+* **What it does**: When set to `true`, the `authenticationInterceptor` will automatically treat all incoming
```
Refer to the `authenticationInterceptor` file.

Slartibartfass2 replied:
Make it a GitHub URL like the others so that it ages better.

Slartibartfass2 replied:
`[AuthenticationInterceptor](https://github.com/SE-UUlm/snowballr-backend/blob/develop/src/main/kotlin/se/uulm/snowballr/backend/grpc/interceptor/AuthenticationInterceptor.kt)` should be fine.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/env/EnvReader.kt:
```diff
@@ -1,49 +1,126 @@
+        // If AUTH_BYPASS_ENABLED is `true`, we must also seed the user
+        val authBypassEnabled = envService.getBooleanOrDefault(AUTH_BYPASS_ENABLED, defaults.authBypassEnabled)
+        val seedUserEnabled =
+            authBypassEnabled || envService.getBooleanOrDefault(DATABASE_SEED_USER_ENABLED, defaults.seedUserEnabled)
```
This env variable (`DATABASE_SEED_USER_ENABLED`) is not in the documentation.

mowi12 replied:
You're right. The `DATABASE_` prefix is missing.

Slartibartfass2 reviewed (changes requested):
almost done

mowi12 reviewed (commented):
Hopefully this should be it!
